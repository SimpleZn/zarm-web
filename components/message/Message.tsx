import React from 'react';
import classnames from 'classnames';
import Icon from '../icon';
import Loading from '../loading';
import { MessageProps, MessageIcon } from './PropsType';
import { mapToIconType, mapToIconTheme, getStyle } from '../notification/utils';

function getIcon(icon: MessageIcon, className: string) {
  if (icon === 'loading') {
    return (
      <Loading className={className} visible size="sm" />
    );
  }
  const iconType = mapToIconType(icon);
  return iconType
    ? (
      <Icon
        type={mapToIconType(icon)}
        className={className}
        size="sm"
        theme={mapToIconTheme(icon)}
      />
    )
    : null;
}

export default class Message extends React.Component<MessageProps, {}> {
  static defaultProps: MessageProps = {
    prefixCls: 'zw-message',
    content: '',
  };

  render() {
    const {
      style, className, prefixCls, icon, content,
      top, bottom,
      onMouseEnter, onMouseLeave, onClick,
    } = this.props;
    const iconToRender = icon ? getIcon(icon, `${prefixCls}__icon`) : null;
    const cls = classnames(`${prefixCls}__content`, { 'has-icon': iconToRender });
    return (
      <div
        className={classnames(prefixCls, className, {
          [`${prefixCls}--fixed`]: top || bottom,
        })}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        style={getStyle(style, top, bottom)}
      >
        <div className={cls} onClick={onClick}>
          {iconToRender}
          <div>{content}</div>
        </div>
      </div>
    );
  }
}
