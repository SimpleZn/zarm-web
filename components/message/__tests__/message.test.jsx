import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Message from '../Message';
import Loading from '../../loading';

describe('Message', () => {
  const content = 'This is a test messgae';
  const { prefixCls } = Message.defaultProps;

  it('renders basic Message correctly', () => {
    const wrapper = shallow(<Message content={content} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text().indexOf(content)).not.toBe(-1);
  });

  it('trigger event correctly', () => {
    const mouseEnter = jest.fn();
    const mouseLeave = jest.fn();
    const click = jest.fn();
    const wrapper = mount(
      <Message
        content={content}
        onMouseEnter={mouseEnter}
        onMouseLeave={mouseLeave}
        onClick={click}
      />,
    );
    wrapper
      .find('div')
      .at(0)
      .simulate('mouseenter')
      .simulate('mouseleave');
    wrapper.find(`.${prefixCls}__content`).simulate('click');
    expect(mouseEnter).toHaveBeenCalledTimes(1);
    expect(mouseLeave).toHaveBeenCalledTimes(1);
    expect(click).toHaveBeenCalledTimes(1);
  });

  it('render loading correctly', () => {
    const wrapper = mount(<Message content={content} icon="loading" />);
    expect(wrapper.find(`.${Loading.defaultProps.prefixCls}`)).toHaveLength(1);
  });

  it('render unknow icon correctly', () => {
    const wrapper = mount(
      <Message content={content} icon="unknown-icon" />,
    );
    expect(wrapper.find(`.${prefixCls}__icon`)).toHaveLength(0);
  });
});
