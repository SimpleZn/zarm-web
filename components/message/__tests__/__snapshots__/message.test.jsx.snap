// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders basic Message correctly 1`] = `
<div
  className="zw-message"
  style={Object {}}
>
  <div
    className="zw-message__content"
  >
    <div>
      This is a test messgae
    </div>
  </div>
</div>
`;
